dask/array/routines.py

```diff
@@ -929,6 +932,304 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
+            "and chunk structure along the first dimension."
+        )
+
+    # if bins is a list, tuple, or array_like then make sure the
+    # length is the same as the number dimensions.
+    if isinstance(bins, (list, tuple)) or is_arraylike(bins):
+        if len(bins) != D:
+            raise ValueError(
+                "The dimension of bins must be equal to the dimension of the sample."
+            )
+    else:
+        # if not then bins must be a single scalar (possibly delayed),
+        # we expand it to the total number of dimensions.
+        bins = (bins,) * D
+
+    # if range is defined, check that it's the right length and also a
+    # sequence of pairs.
+    if range is not None:
+        if len(range) != D:
+            raise ValueError(
+                "range argument requires one entry, a min max pair, per dimension."
+            )
+        if not all(len(r) == 2 for r in range):
+            raise ValueError("range argument should be a sequence of pairs")
+
+    dask_coll_in_bins = any([is_dask_collection(b) for b in bins])
+    dask_coll_in_range = (
+        any([is_dask_collection(r) for r in range]) if range is not None else False
+    )
+
+    # No dask collections passed to bins=... or range=...
+    if not dask_coll_in_bins and not dask_coll_in_range:
+        # We have integer bins and tuple ranges
+        if all(isinstance(b, int) for b in bins) and all(len(r) == 2 for r in range):
+            edges = [np.linspace(r[0], r[1], b + 1) for b, r in zip(bins, range)]
+        # We have arrays of bin edges
+        else:
+            edges = [np.asarray(b) for b in bins]
+        # at this point we've converted bins + range combinations to
+        # edge arrays, so we can null the range argument later passed
+        # to the callable in the graph (_block_histogramdd).
+        range = (None,) * D
+    else:
+        raise NotImplementedError(
+            "Passing dask collections to bins=... or range=... is not supported."
+        )
+
+    # This call is for potential compatibility with delayed bin
+    # configurations. For non-delayed bin confurations this returns
+    # the concrete objects and deps is an empty tuple.
+    (bins_ref, ranges_ref), deps = unpack_collections([edges, range])
+
+    # With dsk below, we will construct a (D + 1) dimensional array
+    # stacked for each chunk. For example, if the histogram is going
+    # to be 3 dimensions, this creates a stack of cubes (1 cube for
+    # each sample chunk) that will be collapsed into a final cube (the
+    # result).
+
+    # This tuple of zeros represents the chunk index along the columns
+    # (we only allow chunking along the rows).
+    column_zeros = tuple(0 for _ in _range(D))
+
+    if weights is None:
+        dsk = {
+            (name, i, *column_zeros): (_block_histogramdd, k, bins_ref, ranges_ref)
+            for i, k in enumerate(flatten(sample.__dask_keys__()))
+        }
```

Comment:
Would it be possible to compute histograms on each chunk and aggregate them? Perhaps using `map_blocks`? Maybe even with [a tree reduction]( https://github.com/dask/dask/blob/707af9dd8b57e1e4235f4f442b36e370f785e171/dask/array/reductions.py#L49 )?